src/main.cpp

```diff
@@ -2062,6 +2062,8 @@ bool DisconnectBlock(const CBlock& block, CValidationState& state, const CBlockI
         }
     }
 
+    // restore the previous block-final transaction hash
+    view.SetFinalTx(blockUndo.final_tx);
```

Comment:
In db474c6f854190e215a4cae418ce675897e5e92a

Do you not need to check `if (enforce_block_final)` here? What if this is the initial block-final block? Does both cases result in the same `blockUndo.final_tx`?

Reply:
I think this code is okay. `final_tx` is initialized to zero if it is not present in the serialized representation. So if we're rewinding past the initial block, then `final_tx` will be cleared, the same as if it had never existed in the first place.

That said, I wonder what the effect on the UTXO database would be? I don't think I delete the key in this instance, so it's not truly identical to earlier state (where the key didn't exist at all and was implicitly zero).

I'll audit that when I'm back at a development machine.

Reply:
So actually there might be some reorg risk around activation here. The line you've highlighted here sets the `final_tx` to zero / null when you reorg past activation. However the txdb.cpp logic does not commit this clearing of the field to the UTXO database.

However, this is not catastrophic as BIP9 comes to the rescue: activation is delayed 2016 blocks so unless you reorg'd more than that, the next time you hit that height the field will be overwritten again with the new initial value.

Still, I'm a little uneasy having that field sitting around in the database with a value that doesn't correspond to any UTXO because that block had been reorg'd out. I'm going to add a special case to delete the record when performing a reorg past the initial block-final coinbase.

Reply:
Done.

Reply:
For the line that this discussion is referencing, yes. I added an RPC test that invalidates a block, then mines essentially the same block with its hash tweaked. This should be fine, but would fail if we were not resetting the FinalTx field on a reorg. It works.

What is not so easy to test is whether the field is deleted if you reorg past activation. This is not visible in any way whatsoever, and even if the field were still there it shouldn't be an issue.